src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureConfig.kt

```diff
@@ -0,0 +1,143 @@
+    val function: RsFunction,
+    var name: String = "",
+    parameters: List<Parameter>,
+    var returnType: Ty,
+    var visibility: RsVis? = null,
+    var isAsync: Boolean = false,
+    var isUnsafe: Boolean = false
+) {
+    val originalParameters = parameters.toList()
+    val parameters = parameters.toMutableList()
+
+    private val parametersText: String
+        get() = parameters.joinToString(", ") { "${it.patText}: ${it.type.renderInsertionSafe()}" }
+
+    fun signature(): String = buildString {
+        visibility?.let { append("${it.text} ") }
+
+        if (isAsync) {
+            append("async ")
+        }
+        if (isUnsafe) {
+            append("unsafe ")
+        }
+        append("fn $name$typeParametersText($parametersText)")
+        if (returnType !is TyUnit) {
+            append(" -> ${returnType.renderInsertionSafe()}")
+        }
+        append(whereClausesText)
+    }
+
+    private val typeParametersText: String
+        get() {
+            val typeParams = typeParameters()
+            if (typeParams.isEmpty()) return ""
+            return typeParams.joinToString(separator = ",", prefix = "<", postfix = ">") { it.text }
+        }
+
+    private val whereClausesText: String
+        get() {
+            val wherePredList = function.whereClause?.wherePredList ?: return ""
+            if (wherePredList.isEmpty()) return ""
+            val typeParams = typeParameters().map { it.declaredType }
+            if (typeParams.isEmpty()) return ""
+            val filtered = wherePredList.filter { it.typeReference?.type in typeParams }
+            if (filtered.isEmpty()) return ""
+            return filtered.joinToString(separator = ",", prefix = " where ") { it.text }
+        }
+
+    private fun typeParameters(): List<RsTypeParameter> {
+        val bounds = typeParameterBounds()
+        val paramAndReturnTypes = mutableSetOf<Ty>()
+        (parameters.map { it.type } + listOfNotNull(returnType)).forEach {
+            paramAndReturnTypes.addAll(it.types())
+            paramAndReturnTypes.addAll(it.dependTypes(bounds))
+        }
+        return function.typeParameters.filter { it.declaredType in paramAndReturnTypes }
+    }
+
+    private fun typeParameterBounds(): Map<Ty, Set<Ty>> =
+        function.typeParameters.associate { typeParameter ->
+            val type = typeParameter.declaredType
+            val bounds = mutableSetOf<Ty>()
+            typeParameter.bounds.flatMapTo(bounds) {
+                it.bound.traitRef?.path?.typeArguments?.flatMap { it.type.types() }.orEmpty()
+            }
+            type to bounds
+        }
+
+    fun createChangeInfo(): ChangeInfo = RsSignatureChangeInfo(this)
+
+    companion object {
+        fun create(function: RsFunction): RsFunctionSignatureConfig {
+            return RsFunctionSignatureConfig(function,
+                function.name ?: "",
+                function.valueParameters.mapIndexed { index, it ->
```

Comment:
I wanted to do that, but I wonder how to provide a better error message? When I use `ChangeSignatureHandler`, it only has `findTargetMember`, which can return null, and in that case  `getTargetNotFoundMessage` is displayed.

So it would need to have a generic error message, something like `Must target a function with non-cfg disabled parameters`. Is it ok or is there a better way?

Reply:
You can check it e.g. in `showRefactoringDialog` and if necessary show error message using `CommonRefactoringUtil.showErrorHint` (and not show dialog). Python [does this](https://github.com/JetBrains/intellij-community/blob/7d654ffc0eaee2129850b583112f235792671e75/python/src/com/jetbrains/python/refactoring/changeSignature/PyChangeSignatureHandler.java#L91-L103)